### PR TITLE
[IMP] mrp: Mass assignation of SN/Lot in MRP

### PR DESCRIPTION
**Pull request by ajf-odoo**
Allow to produce several serial numbers at once.
Only when BoM do not contain any component tracked by unique serial number
and lot ones are from 1 lot.
Also allow to copy/paste serial numbers.
This applies to manufacturing orders and subcontracting receipts, dropshipping handled.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/wizard/mrp_serial_mass_produce.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import Counter
+
+from odoo.exceptions import UserError
+from odoo import _, api, fields, models
+
+
+class MrpSerialMassProduce(models.TransientModel):
```
Comment (amoyaux) [inline addons/mrp/wizard/mrp_serial_mass_produce.py:10]: Override stock.assign.serial

File: addons/mrp/wizard/mrp_serial_mass_produce.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            self.serial_numbers = "\n".join([self.serial_numbers, generated_serial_numbers]) if self.serial_numbers else generated_serial_numbers
+            self._onchange_serial_numbers()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': self._description,
+            'res_model': self._name,
+            'res_id': self.id,
+            'view_mode': 'form',
+            'target': 'new',
+        }
```
Comment (amoyaux) [inline addons/mrp/wizard/mrp_serial_mass_produce.py:35]: Put that in a xml

**Comment by amoyaux** [comment]
@ajf-odoo C'est surtout de la forme pour la majeur partie, ça semble bon. Le seul problème c'est la génération des backorders. Il faut tout créer d'un seul coup. On peut voir ça ensemble au bureau si tu veux.

File: addons/mrp_subcontracting_dropshipping/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    @api.model
+    def create_missing_subcontracting_dropshipping_rules(self):
+        route = self.env.ref('mrp_subcontracting_dropshipping.route_subcontracting_dropshipping')
+        company_ids = self.env['res.company'].search([])
+        company_has_rules = self.env['stock.rule'].search([('route_id', '=', route.id)]).mapped('company_id')
+        company_todo_rules = company_ids - company_has_rules
+        company_todo_rules._create_subcontracting_dropshipping_rules()
+
+    def _create_per_company_rules(self):
+        super()._create_per_company_rules()
```
Comment (amoyaux) [inline addons/mrp_subcontracting_dropshipping/models/res_company.py:43]: always return super value at then end of functione

File: addons/mrp_subcontracting_dropshipping/models/res_company.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        supplier_location = self.env.ref('stock.stock_location_suppliers')
+        vals = []
+        for company in self:
+            subcontracting_location = company.subcontracting_location_id
+            dropship_picking_type = self.env['stock.picking.type'].search([
+                ('company_id', '=', company.id),
+                ('default_location_src_id.usage', '=', 'supplier'),
+                ('default_location_dest_id.usage', '=', 'customer'),
+            ], limit=1, order='sequence')
+            vals.append({
```
Comment (amoyaux) [inline addons/mrp_subcontracting_dropshipping/models/res_company.py:21]: if not dropship_picking_type:
    return False

File: addons/mrp/models/mrp_production.py [hunk 5]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+    def action_serial_mass_produce_wizard(self):
+        self.ensure_one()
+        self._check_company()
+        if self.state != 'confirmed':
+            return
+        if self.product_id.tracking != 'serial':
+            return
+        have_serial_components, missing_components, multiple_lot_components = self._check_serial_mass_produce_components()
+        if have_serial_components:
+            raise UserError(_("Can't work with components tracked by serial number"))
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:1791]: Is it possible to have this error from the UI? If not it's probably a better idea to raise a not implemented error for serial numbers

File: addons/purchase_mrp/models/purchase.py [hunk 6]
```diff
@@ -18,12 +18,18 @@ class PurchaseOrder(models.Model):
...
         for purchase in self:
-            purchase.mrp_production_count = len(purchase.order_line.move_dest_ids.group_id.mrp_production_ids |
-                                                purchase.order_line.move_ids.move_dest_ids.group_id.mrp_production_ids)
+            purchase.mrp_production_count = len(purchase._get_mrp_productions())
+
+    def _get_mrp_productions(self, allow_archived_picking_types=False):
+        productions = self.order_line.move_dest_ids.group_id.mrp_production_ids
+        productions |= self.order_line.move_ids.move_dest_ids.group_id.mrp_production_ids
+        if not allow_archived_picking_types:
+            productions = productions.filtered(lambda p: p.picking_type_id.active)
```
Comment (Whenrow) [inline addons/purchase_mrp/models/purchase.py:27]: It seems all this code in purchase should be moved in the dropship commit

**Review by Whenrow — COMMENTED** [review]
Here are the first comments I have on your task

File: addons/stock/models/stock_production_lot.py [hunk 7]
```diff
@@ -27,6 +30,47 @@ class ProductionLot(models.Model):
...
+
+    @api.model
+    def get_next_serial(self, company, product):
+        """ This method will return the next serial number to be attributed to the product
+        based on the highest generated.
+        """
+        if product.tracking == "serial":
+            last_serial = self.env['stock.production.lot'].search(
+                [('company_id', '=', company.id), ('product_id', '=', product.id)],
+                limit=1, order='name DESC')
```
[the remark this answers is not in the document]
Comment (ticodoo) [reply addons/stock/models/stock_production_lot.py:73]: This definitely needs to be changed. If you use order='name DESC' you will have issues with repeat SNs by default once you have cases such as:
SNs used = 'a9', 'a10' => 'a9' = last serial and 'a10' will always the default "First SN" in your wizard.

File: addons/mrp_subcontracting/models/stock_move.py [hunk 8]
```diff
@@ -151,6 +163,48 @@ def _action_record_components(self):
...
+            message = _("Make sure enough quantities of these components are reserved to carry on production:\n")
+            message += "\n".join(component.name for component in missing_components)
+            pickings = self.env['stock.picking'].search([
+                ('group_id', '=', production.procurement_group_id.id),
+                ('state', 'not in', ('done', 'cancel'))
+            ])
+            purchase_orders = self.env['purchase.order'].search([
+                ('group_id', '=', production.procurement_group_id.id),
+                ('state', 'not in', ('purchase', 'done', 'cancel'))
+            ])
```
Comment (ticodoo) [inline addons/mrp_subcontracting/models/stock_move.py:180]: Subcontracting does not depend on the purchase module and a PO is optional for subcontracting (can be done directly with a receipt Receive From = subcontractor, Product = subcontracted product), so this will throw a server error when user clicks on button without purchase installed

File: addons/mrp_subcontracting/models/stock_move.py [hunk 9]
```diff
@@ -25,6 +26,17 @@ def _compute_show_subcontracting_details_visible(self):
...
 
+    def _compute_show_serial_mass_receipt(self):
+        """ Compute if the action button for serial mass receipt is visible """
+        self.show_serial_mass_receipt = False
+        for move in self:
+            if move.state not in ('draft', 'cancel', 'done') and move.is_subcontract and move.product_id.tracking == "serial" \
+                    and float_compare(move.product_uom_qty, 1, precision_rounding=move.product_uom.rounding) > 0:
+                production = move.move_orig_ids.production_id
+                have_serial_components, have_lot_components, dummy, multiple_lot_components = production._check_serial_mass_produce_components()
+                if not have_serial_components and have_lot_components and not multiple_lot_components:
```
Comment (ticodoo) [inline addons/mrp_subcontracting/models/stock_move.py:37]: Is there a reason why subcontracting requires at least one of its components to be lot tracked? I would think you could mass assign SNs even if none of its components are tracked, but maybe I am missing something?

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 10]
```diff
@@ -36,11 +36,16 @@ def _action_done(self):
         for move in self.move_lines.filtered(lambda move: move.is_subcontract):
             # Auto set qty_producing/lot_producing_id of MO if there isn't tracked component
             # If there is tracked component, the flow use subcontracting_record_component instead
```
Comment (ticodoo) [inline addons/mrp_subcontracting/models/stock_picking.py:38]: You should update/delete this comment if you're going to delete the code it is related to

File: addons/mrp_subcontracting/views/stock_picking_views.xml [hunk 11]
```diff
@@ -10,9 +10,17 @@
...
+                <button name="action_serial_mass_receipt" type="object" icon="fa-list" width="0.1"
+                    string="Mass receipt"
+                    attrs="{'invisible': [('show_serial_mass_receipt', '=', False)]}" options='{"warn": true}'/>
                 <field name="show_subcontracting_details_visible" invisible="1"/>
                 <button name="action_show_subcontract_details" string="Register components for subcontracted product" type="object" icon="fa-sitemap"
                     width="0.1" attrs="{'invisible': [('show_subcontracting_details_visible', '=', False)]}"/>
+                <field name="is_subcontract" invisible="1"/>
+            </xpath>
+            <xpath expr="//field[@name='move_ids_without_package']//tree//button[@name='action_assign_serial']" position="attributes">
+                <attribute name="attrs">{'invisible': ['|', '|', ('display_assign_serial', '=', False), ('show_operations', '=', False), ('is_subcontract', '=', True)]}</attribute>
```
Comment (ticodoo) [inline addons/mrp_subcontracting/views/stock_picking_views.xml:23]: I think something isn't quite correct here? 

Odoo-WH-IN-00035

First burger:
Screenshot from 2021-08-30 14-46-42

Second burger:
Screenshot from 2021-08-30 14-47-06

I didn't look closely as to why this is happening, but the first burger is showing `move_line_ids` while the second burger is showing `move_line_nosuggest_ids`. I have Storage Locations active if that helps.

File: addons/mrp_subcontracting_dropshipping/models/purchase.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, _
+from odoo.exceptions import UserError
+
+
+class PurchaseOrder(models.Model):
```
Comment (ticodoo) [inline addons/mrp_subcontracting_dropshipping/models/purchase.py:8]: These purchase/dropshipping changes again seems separate from the mass SN assigning and should be in a separate commit. so it's clear what it's purpose is.

**Comment by ticodoo** [comment]
The rest looks good to me, nice job! In addition to the changes/fixes you should also clean up your commit messages a bit, e.g.:

- For your fix commit, it's best if you can write the steps to reproduce the bug you're fixing (in this case you can just write it for one of the 3 fixes since it's basically the same flow to reproduce the bug in each case)
- for your mass assign SN commits, you don't need to repeat the module name in the title since it's implied (i.e. `[IMP] mrp_subcontracting: Mass assignation of SN/Lot` implies it's for mrp_subcontracting)

File: addons/mrp/models/mrp_production.py [hunk 13]
```diff
@@ -522,7 +522,7 @@ def _compute_forecasted_issue(self):
             if order.product_id:
                 virtual_available = order.product_id.with_context(warehouse=warehouse.id, to_date=order.date_planned_start).virtual_available
                 if order.state == 'draft':
-                    virtual_available += order.product_uom_qty
+                    virtual_available += order.product_qty
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py:556]: new pr 74961
& specific commit in master

File: addons/mrp/models/mrp_production.py [hunk 14]
```diff
@@ -509,6 +510,17 @@ def _compute_show_lot_ids(self):
         for order in self:
             order.show_lot_ids = order.state != 'draft' and any(m.product_id.tracking == 'serial' for m in order.move_raw_ids)
 
+    @api.depends('state', 'move_raw_ids')
+    def _compute_show_serial_mass_produce(self):
+        for order in self:
+            order.show_serial_mass_produce = False
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: done

File: addons/mrp/models/mrp_production.py [hunk 15]
```diff
@@ -1392,20 +1404,28 @@ def _generate_backorder_productions(self, close_mo=True):
...
                         else:
                             move_vals[0]['production_id'] = backorder_mo.id
-                        new_moves_vals.append(move_vals[0])
-                new_moves = self.env['stock.move'].create(new_moves_vals)
+                        new_move = self.env['stock.move'].create(move_vals[0])
+                        if self.env.context.get('serial_mass_produce') and move.raw_material_production_id:
+                            # copy & adapt move line for raw materials in backorder
+                            move.with_context(bypass_reservation_update=True).move_line_ids[0].copy({
+                                'move_id': new_move.id,
+                                'product_uom_qty': uom_qty_to_split,
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: ok

File: addons/mrp/models/mrp_production.py [hunk 16]
```diff
@@ -1392,20 +1404,28 @@ def _generate_backorder_productions(self, close_mo=True):
...
+                                'lot_id': move.move_line_ids[0].lot_id.id if move.move_line_ids[0].lot_id else 0,
+                            })
+                            # adapt original move
+                            move.with_context(bypass_reservation_update=True).move_line_ids[0].product_uom_qty = move.product_uom_qty
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: ok

File: addons/mrp/models/mrp_production.py [hunk 17]
```diff
@@ -1746,3 +1766,74 @@ def _check_immediate(self):
...
+            raise UserError(_('Expected product tracked by Serial Number.'))
+        if cancel_remaining_quantities is None and self.product_qty != len(serial_numbers):
+            raise UserError(_('Serial Numbers mismatch or missing cancel_remaining_quantities flag.'))
+        result = []
+        last_serial_number = serial_numbers[-1]
+        production = self
+        for serial_number in serial_numbers:
+            production = production.with_context(serial_mass_produce=True)
+            production.qty_producing = 1
+            backorder = production._generate_backorder_productions(False) if production.product_qty > 1 and (serial_number != last_serial_number or not cancel_remaining_quantities) else None
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: pending, a v2 task has been created

File: addons/mrp/models/stock_move.py [hunk 18]
```diff
@@ -327,6 +327,8 @@ def _should_bypass_set_qty_producing(self):
         if float_is_zero(self.product_uom_qty, precision_rounding=self.product_uom.rounding):
             return True
         if self.has_tracking != 'none' or self.state == 'done':
+            if self.env.context.get('serial_mass_produce'):
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/stock_move.py]: done

File: addons/mrp/views/mrp_production_views.xml [hunk 19]
```diff
@@ -575,7 +577,18 @@
...
+            <field name="name">Serial Mass Produce</field>
+            <field name="model_id" ref="mrp.model_mrp_production"/>
+            <field name="binding_model_id" ref="mrp.model_mrp_production"/>
+            <field name="binding_view_types">form</field>
+            <field name="state">code</field>
+            <field name="code">
+                action = record.action_serial_mass_produce_wizard()
+            </field>
+        </record>
+-->
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/views/mrp_production_views.xml]: done

File: addons/mrp/wizard/mrp_serial_mass_produce.py (hunk 1 above)
Comment (ajf-odoo) [reply addons/mrp/wizard/mrp_serial_mass_produce.py]: done

File: addons/mrp/wizard/mrp_serial_mass_produce.py (hunk 2 above)
Comment (ajf-odoo) [reply addons/mrp/wizard/mrp_serial_mass_produce.py]: done

File: addons/mrp_subcontracting_dropshipping/models/res_company.py (hunk 3 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting_dropshipping/models/res_company.py]: ok, done

File: addons/mrp_subcontracting_dropshipping/models/res_company.py (hunk 4 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting_dropshipping/models/res_company.py]: ok, done

File: addons/mrp/models/mrp_production.py (hunk 5 above)
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: removed, remaining test needed when entry was by action menu

File: addons/mrp/models/mrp_production.py [hunk 20]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+        }
+        return action
+
+    def _check_serial_mass_produce_components(self):
+        have_serial_components = False
+        missing_components = set()
+        multiple_lot_components = set()
+        for production in self:
+            have_serial_components |= any(move.product_id.tracking == 'serial' for move in production.move_raw_ids)
+            missing_components.update([move.product_id for move in production.move_raw_ids if move.reserved_availability < move.product_uom_qty])
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: done

File: addons/mrp/models/mrp_production.py [hunk 21]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+            components = {}
+            for move in production.move_raw_ids:
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py:1870]: checked with arm

File: addons/mrp/models/mrp_production.py [hunk 22]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+            'default_next_serial_number': next_serial,
+        }
+        return action
+
+    def _check_serial_mass_produce_components(self):
+        have_serial_components = False
+        missing_components = set()
+        multiple_lot_components = set()
+        for production in self:
+            have_serial_components |= any(move.product_id.tracking == 'serial' for move in production.move_raw_ids)
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py:1866]: idem

File: addons/mrp/models/mrp_production.py [hunk 23]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+    def _generate_backorder_productions_multi(self, serial_numbers, cancel_remaining_quantities=False):
+        self.ensure_one()
+        if self.product_id.tracking != 'serial':
+            raise UserError(_('Expected product tracked by Serial Number.'))
+        if self.backorder_sequence == 0:
+            self.backorder_sequence = 1
+        result = []
+        production = self
+        for serial_number in serial_numbers:
+            production.qty_producing = 1
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py:1889]: checked with arm

File: addons/purchase_mrp/models/purchase.py (hunk 6 above)
Comment (ajf-odoo) [reply addons/purchase_mrp/models/purchase.py]: commits adapted

File: addons/mrp/models/mrp_production.py [hunk 24]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
...
+        if self.state != 'confirmed':
+            return
+        if self.product_id.tracking != 'serial':
+            return
+        have_serial_components, missing_components, multiple_lot_components = self._check_serial_mass_produce_components()
+        if have_serial_components:
+            raise UserError(_("Can't work with components tracked by serial number"))
+        message = ""
+        if missing_components:
+            message += _("Make sure enough quantities of these components are reserved to carry on production:") + "\n"
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: ok

File: addons/stock/models/stock_production_lot.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/stock/models/stock_production_lot.py:74]: Considering generate_lot_names and user documentation showing 'SN-001' as first sn in sequence to be generated,
ordering by name seemed the best solution for real-life sn (fixed length with zero padding for the sn part).
-> checked with mgm and confirmed this way 
Set the very first sn to 'a0001' rather than 'a1' to solve this issue.
However, I adapted the code to handle full sequences or sequence like 1,2,...,9,10,...

File: addons/stock/models/stock_production_lot.py [hunk 25]
```diff
@@ -27,6 +30,47 @@ class ProductionLot(models.Model):
...
+            if last_serial:
+                return self.env['stock.production.lot'].generate_lot_names(last_serial.name, 2)[1]
+        return ""
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/stock/models/stock_production_lot.py]: done

File: addons/mrp/models/mrp_production.py [hunk 26]
```diff
@@ -1766,3 +1778,142 @@ def _check_immediate(self):
                     ) and float_is_zero(production.qty_producing, precision_digits=pd):
                 immediate_productions |= production
         return immediate_productions
+
+    def action_serial_mass_produce_wizard(self):
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp/models/mrp_production.py]: corrected

File: addons/purchase_mrp/models/purchase.py [hunk 27]
```diff
@@ -18,12 +18,18 @@ class PurchaseOrder(models.Model):
     @api.depends('order_line.move_dest_ids.group_id.mrp_production_ids')
     def _compute_mrp_production_count(self):
         for purchase in self:
-            purchase.mrp_production_count = len(purchase.order_line.move_dest_ids.group_id.mrp_production_ids |
-                                                purchase.order_line.move_ids.move_dest_ids.group_id.mrp_production_ids)
+            purchase.mrp_production_count = len(purchase._get_mrp_productions())
+
+    def _get_mrp_productions(self, allow_archived_picking_types=False):
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/purchase_mrp/models/purchase.py]: no, code removed

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 28]
```diff
@@ -54,7 +59,7 @@ def _action_done(self):
                 if move_line.lot_id:
                     production.lot_producing_id = move_line.lot_id
                 production.qty_producing = move_line.product_uom_id._compute_quantity(move_line.qty_done, production.product_uom_id)
-                production._set_qty_producing()
+                production.with_context(subcontract_move_id=True)._set_qty_producing()
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp_subcontracting/models/stock_picking.py]: done

File: addons/mrp_subcontracting/models/stock_move.py [hunk 29]
```diff
@@ -151,6 +163,48 @@ def _action_record_components(self):
...
+        if self.show_serial_mass_receipt:
+            return super().action_show_details()
+        return res
+
+    def action_clear_lines_show_details(self):
+        res = super().action_clear_lines_show_details()
+        if self.show_serial_mass_receipt:
+            return super().action_show_details()
+        return res
+
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp_subcontracting/models/stock_move.py:190]: needed as record components can return its own action

File: addons/mrp_subcontracting_dropshipping/models/purchase.py [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+        if self.dest_address_id:
+            if self.mrp_production_count:
+                mrp_production_ids = self._get_mrp_productions(True)
+                if self.dest_address_id in mrp_production_ids.bom_id.subcontractor_ids:
+                    return self.dest_address_id.property_stock_subcontractor.id
+            elif not self.sale_order_count:
+                in_bom_products = False
+                not_in_bom_products = False
+                for order_line in self.order_line:
+                    if any(bom_line.bom_id.type == 'subcontract' and self.dest_address_id in bom_line.bom_id.subcontractor_ids for bom_line in order_line.product_id.bom_line_ids):
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/mrp_subcontracting_dropshipping/models/purchase.py]: done, well seen

File: addons/purchase/models/purchase.py [hunk 31]
```diff
@@ -90,7 +90,7 @@ def _compute_invoice(self):
         help="Depicts the date within which the Quotation should be confirmed and converted into a purchase order.")
     date_approve = fields.Datetime('Confirmation Date', readonly=1, index=True, copy=False)
     partner_id = fields.Many2one('res.partner', string='Vendor', required=True, states=READONLY_STATES, change_default=True, tracking=True, domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", help="You can find a vendor by its Name, TIN, Email or Internal Reference.")
-    dest_address_id = fields.Many2one('res.partner', domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", string='Drop Ship Address', states=READONLY_STATES,
+    dest_address_id = fields.Many2one('res.partner', domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", string='Dropship Address', states=READONLY_STATES,
```
[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/purchase/models/purchase.py:93]: commits splitted

File: addons/mrp_subcontracting/models/stock_move.py (hunk 8 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting/models/stock_move.py]: well seen, adapted

File: addons/mrp_subcontracting/models/stock_move.py (hunk 9 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting/models/stock_move.py:37]: Mass Receipt bypasses Record Components by using standard mechanism when no serial or multiple lot component.
If no tracked component at all, another button doing the same serves nothing

File: addons/mrp_subcontracting/models/stock_picking.py (hunk 10 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting/models/stock_picking.py:38]: done

File: addons/mrp_subcontracting/views/stock_picking_views.xml (hunk 11 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting/views/stock_picking_views.xml:23]: this is related to show_reserved (Pre-fill Detailed Operations), standard process

File: addons/mrp_subcontracting_dropshipping/models/purchase.py (hunk 12 above)
Comment (ajf-odoo) [reply addons/mrp_subcontracting_dropshipping/models/purchase.py:8]: commits splitted

File: addons/stock/models/stock_production_lot.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ticodoo) [reply addons/stock/models/stock_production_lot.py:74]: Good point about the real-life sn. I agree it's better have safer though because I'm sure there will be users who will vary the length of their SNs and run into this problem. I'm not sure if I understand your solution though, wouldn't it be easier to assume users will always create the SNs in order so we can always search by create_date to determine the next SN's name? Especially since it looks like we will do 2 searches whenever the last_serial initial_number[0] != '0', which will happen 9 out of 10 times?

File: addons/mrp_subcontracting/views/stock_picking_views.xml (hunk 11 above)
Comment (ticodoo) [reply addons/mrp_subcontracting/views/stock_picking_views.xml:23]: I think it's confusing to have both show what looks to be the same window, but with different info in it. There also seems to be some weird behavior where sometimes the first burger shows the "Detailed Operations" window and other times it shows the "Subcontract" window, specifically if you delete one of the `move_line_ids` lines, close the window and reopen. This is maybe something to consider fixing as a follow up task.

File: addons/mrp/models/mrp_production.py [hunk 32]
```diff
@@ -1952,3 +1952,12 @@ def _generate_backorder_productions_multi(self, serial_numbers, cancel_remaining
...
             production = backorder
         return result
+
+    def _get_missing_components_origins(self):
+        names = []
+        pickings = self.env['stock.picking'].search([
+            ('group_id', '=', self.procurement_group_id.id),
+            ('state', 'not in', ('done', 'cancel'))
+        ])
+        names += [picking.name for picking in pickings]
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:1962]: pickings.mapped('name') does the same

**Comment by Whenrow** [comment]
robodoo rebase-ff
